File: 071_Symmetric Tree.py
"""
Symmetric Tree
Check whether a binary tree is a mirror of itself (symmetric around its center).
"""
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def is_mirror(a: Optional[TreeNode], b: Optional[TreeNode]) -> bool:
    if not a and not b:
        return True
    if not a or not b:
        return False
    return a.val == b.val and is_mirror(a.left, b.right) and is_mirror(a.right, b.left)

def is_symmetric(root: Optional[TreeNode]) -> bool:
    if not root:
        return True
    return is_mirror(root.left, root.right)

def build_example1():
    r = TreeNode(1)
    r.left = TreeNode(2)
    r.right = TreeNode(2)
    r.left.left = TreeNode(3)
    r.left.right = TreeNode(4)
    r.right.left = TreeNode(4)
    r.right.right = TreeNode(3)
    return r

def build_example2():
    r = TreeNode(1)
    r.left = TreeNode(2)
    r.right = TreeNode(2)
    r.left.right = TreeNode(3)
    r.right.right = TreeNode(3)
    return r

if __name__ == '__main__':
    a = build_example1()
    b = build_example2()
    print('Example1 is symmetric:', is_symmetric(a))  # True
    print('Example2 is symmetric:', is_symmetric(b))  # False
    print('Empty tree symmetric:', is_symmetric(None))

# Time Complexity: O(n)
# Space Complexity: O(h) recursion stack where h is tree height

File: 071_Symmetric Tree.cpp
// Symmetric Tree
// Check whether a binary tree is a mirror of itself (symmetric around its center).
// Solution: recursively compare left and right subtrees for mirror property.
#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left, *right;
    TreeNode(int x): val(x), left(nullptr), right(nullptr) {}
};

bool isMirror(TreeNode* a, TreeNode* b) {
    if (!a && !b) return true;
    if (!a || !b) return false;
    return a->val == b->val && isMirror(a->left, b->right) && isMirror(a->right, b->left);
}

bool isSymmetric(TreeNode* root) {
    if (!root) return true;
    return isMirror(root->left, root->right);
}

// small helper to build a tree for tests
TreeNode* buildExample1() { // symmetric
    TreeNode* r = new TreeNode(1);
    r->left = new TreeNode(2);
    r->right = new TreeNode(2);
    r->left->left = new TreeNode(3);
    r->left->right = new TreeNode(4);
    r->right->left = new TreeNode(4);
    r->right->right = new TreeNode(3);
    return r;
}

TreeNode* buildExample2() { // not symmetric
    TreeNode* r = new TreeNode(1);
    r->left = new TreeNode(2);
    r->right = new TreeNode(2);
    r->left->right = new TreeNode(3);
    r->right->right = new TreeNode(3);
    return r;
}

int main() {
    TreeNode* a = buildExample1();
    TreeNode* b = buildExample2();
    cout << boolalpha;
    cout << "Example1 is symmetric: " << isSymmetric(a) << "\n"; // true
    cout << "Example2 is symmetric: " << isSymmetric(b) << "\n"; // false
    cout << "Empty tree symmetric: " << isSymmetric(nullptr) << "\n"; // true
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(h) recursion stack where h is tree height
